Refactor C->C to C->R twice

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56964 Clean up naming and comments
* #55871 Refactor fast gradcheck
* #55699 Enable fixed fast_mode for complex
* **#55692 Refactor C->C to C->R twice**

### Release notes
get_numerical_jacobian and get_analytical_jacobian only support `grad_out=1` and `fn` no longer accepts functions that return complex output

Differential Revision: [D28004614](https://our.internmc.facebook.com/intern/diff/D28004614)

soulitzer committed:
Refactor C->C to C->R twice

[ghstack-poisoned]

soulitzer committed:
Update on "Refactor C->C to C->R twice"




[ghstack-poisoned]

soulitzer committed:
Update on "Refactor C->C to C->R twice"




[ghstack-poisoned]

soulitzer committed:
Update on "Refactor C->C to C->R twice"




[ghstack-poisoned]

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -182,40 +182,25 @@ def compute(a, b):
-            # wirtinger derivative
-            w_d = 0.5 * (ds_dx - ds_dy * 1j)
-            jacobians_cols.append(grad_out.conjugate() * conj_w_d + grad_out * w_d.conj())
+            jacobians_cols.append(2 * grad_out * conj_w_d)
```
You can remove the 0.5 above and the 2 at this line :)

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -1033,19 +1004,16 @@ def gradcheck_helper(func, inputs, eps, atol, rtol, check_sparse_nnz, nondet_tol
-    if fast_mode and has_complex_inputs_or_outputs(tupled_inputs, func_out):
-        raise NotImplementedError("Fast mode for gradcheck and gradgradcheck is currently only implemented"
-                                  " for R to R functions.")
+    # if fast_mode and has_complex_inputs_or_outputs(tupled_inputs, func_out):
```
Should this be deleted?

soulitzer replied:
Whoops

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -1033,19 +1004,16 @@ def gradcheck_helper(func, inputs, eps, atol, rtol, check_sparse_nnz, nondet_tol
-    else:
-        slow_gradcheck(func, func_out, tupled_inputs, outputs, eps, rtol,
-                       atol, check_grad_dtypes, nondet_tol)
+    complex_indices = [i for i, o in enumerate(outputs) if o.is_complex()]
```
Do we actually have any example of a function that returns mixed real and complex outputs? Could we assume they are either all real or all complex?

soulitzer replied:
As discussed offline, keeping this since it is tested and doesn't add too much complexity

albanD reviewed (commented):
I like when a refactor removes a lot more code than it adds :D

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -182,40 +182,25 @@ def compute(a, b):
     # and z = x + yj for complex valued input
     jacobians_cols: List[torch.Tensor] = []
     ds_dx_tup = jvp_fn(delta)
+    assert(not isinstance(grad_out, complex))
```
All the grad_out logic should be removed here right? This function is only ever called with grad_out=1 IIUC

soulitzer committed:
Update on "Refactor C->C to C->R twice"




[ghstack-poisoned]

soulitzer committed:
Update on "Refactor C->C to C->R twice"




[ghstack-poisoned]

soulitzer committed:
Update on "Refactor C->C to C->R twice"


### Release notes
get_numerical_jacobian and get_analytical_jacobian no longer accept the `grad_out` parameter, and `fn` cannot return complex output


[ghstack-poisoned]

soulitzer committed:
Update on "Refactor C->C to C->R twice"


### Release notes
get_numerical_jacobian and get_analytical_jacobian only support `grad_out=1` and `fn` no longer accepts functions that return complex output


[ghstack-poisoned]

soulitzer committed:
Update on "Refactor C->C to C->R twice"


### Release notes
get_numerical_jacobian and get_analytical_jacobian only support `grad_out=1` and `fn` no longer accepts functions that return complex output


[ghstack-poisoned]

soulitzer committed:
Update on "Refactor C->C to C->R twice"


### Release notes
get_numerical_jacobian and get_analytical_jacobian only support `grad_out=1` and `fn` no longer accepts functions that return complex output


[ghstack-poisoned]

soulitzer committed:
Update on "Refactor C->C to C->R twice"


### Release notes
get_numerical_jacobian and get_analytical_jacobian only support `grad_out=1` and `fn` no longer accepts functions that return complex output


[ghstack-poisoned]

soulitzer committed:
Update on "Refactor C->C to C->R twice"


### Release notes
get_numerical_jacobian and get_analytical_jacobian only support `grad_out=1` and `fn` no longer accepts functions that return complex output


[ghstack-poisoned]

soulitzer committed:
Update on "Refactor C->C to C->R twice"


### Release notes
get_numerical_jacobian and get_analytical_jacobian only support `grad_out=1` and `fn` no longer accepts functions that return complex output


[ghstack-poisoned]

soulitzer committed:
Update on "Refactor C->C to C->R twice"


### Release notes
get_numerical_jacobian and get_analytical_jacobian only support `grad_out=1` and `fn` no longer accepts functions that return complex output


[ghstack-poisoned]

soulitzer committed:
Update on "Refactor C->C to C->R twice"


### Release notes
get_numerical_jacobian and get_analytical_jacobian only support `grad_out=1` and `fn` no longer accepts functions that return complex output


[ghstack-poisoned]

soulitzer committed:
Update on "Refactor C->C to C->R twice"


### Release notes
get_numerical_jacobian and get_analytical_jacobian only support `grad_out=1` and `fn` no longer accepts functions that return complex output


[ghstack-poisoned]

soulitzer committed:
Update on "Refactor C->C to C->R twice"


### Release notes
get_numerical_jacobian and get_analytical_jacobian only support `grad_out=1` and `fn` no longer accepts functions that return complex output


[ghstack-poisoned]

soulitzer committed:
Update on "Refactor C->C to C->R twice"


### Release notes
get_numerical_jacobian and get_analytical_jacobian only support `grad_out=1` and `fn` no longer accepts functions that return complex output


[ghstack-poisoned]

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -162,10 +162,13 @@ def get_numerical_jacobian(fn, inputs, target=None, eps=1e-3, grad_out=1.0):
                   "in a future version of PyTorch. If you have a specific use for "
                   "this or feature request for this to be a stable API, please file "
                   "us an issue at https://github.com/pytorch/pytorch/issues/new")
+    if grad_out != 1.0:
```
nit: could you add a comment here saying that the parameter is only kept for BC reasons

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -424,15 +408,21 @@ def get_analytical_jacobian(inputs, output, nondet_tol=0.0, grad_out=1.0):
                   "in a future version of PyTorch. If you have a specific use for "
                   "this or feature request for this to be a stable API, please file "
                   "us an issue at https://github.com/pytorch/pytorch/issues/new")
+    if grad_out != 1.0:
```
Same comment about BC

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -961,8 +938,8 @@ def gradcheck(
     computation is done under the assumption that the overall function has a real valued output.
-    For functions with complex output, gradcheck compares the numerical and analytical gradients
-    for two values of :attr:`grad_output`: 1 and 1j. For more details, check out
+    For functions with complex output, gradcheck treats them as if they are two separate functions
```
Maybe we can phrase this better?
gradcheck is applied to two real-valued functions corresponding to taking the real part of the complex outputs  for the first one and the imaginary part of the complex outputs for the second one.

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -725,47 +716,65 @@ def _differentiable_outputs(x):
+    out_is_imag = complex_indices and output_idx in complex_indices
+    grad_output = "1j" if out_is_imag else "1"
+    prefix = "" if not inp_is_complex and not out_is_imag else \
+        f"Gradients failed to compare equal for grad output = {grad_output}"
```
Well, we shouldn't talk about grad_output here anymore.
Maybe only "while considering the real part of the complex outputs only"

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -725,47 +716,65 @@ def _differentiable_outputs(x):
+        real_func_out = real_fn(*tupled_inputs)
+        real_outputs = _differentiable_outputs(real_func_out)
+        gradcheck_fn(real_fn, real_func_out, tupled_inputs, real_outputs, eps,
+                     rtol, atol, check_grad_dtypes, nondet_tol)
```
You're missing complex_indices no?

soulitzer replied:
Good catch!

soulitzer replied:
This isn't being tested though, which is why we did not catch this. We'd need a function that fails while considering the real part of a complex output, but succeeds when considering only the imaginary part, or maybe there is some way to accomplish this with hooks.

albanD replied:
Yes I think you can add some hooks on the result and add noise to either the real or imaginary part of the gradients?

albanD reviewed (commented):
LGTM

Could you add one test to make sure that the right error is raised?

soulitzer committed:
Update on "Refactor C->C to C->R twice"


### Release notes
get_numerical_jacobian and get_analytical_jacobian only support `grad_out=1` and `fn` no longer accepts functions that return complex output


[ghstack-poisoned]

soulitzer committed:
Update on "Refactor C->C to C->R twice"


### Release notes
get_numerical_jacobian and get_analytical_jacobian only support `grad_out=1` and `fn` no longer accepts functions that return complex output


[ghstack-poisoned]

soulitzer commented on test/test_autograd.py:
```diff
@@ -4246,10 +4246,13 @@ def fn(inputs):
         self.assertEqual(jacobian[0], 2 * torch.eye(4))
         self.assertEqual(jacobian[1], 1 * torch.eye(4))
 
+        with self.assertRaisesRegex(ValueError, "Expected grad_out to be 1.0"):
```
New test added to make sure right error is raised

albanD reviewed (approved):
Thanks for the update! Looks good!

You will need to fix the lint though.

soulitzer committed:
Update on "Refactor C->C to C->R twice"


### Release notes
get_numerical_jacobian and get_analytical_jacobian only support `grad_out=1` and `fn` no longer accepts functions that return complex output


[ghstack-poisoned]
